src/lib/config.ts

```diff
@@ -0,0 +1,417 @@
+// Manages ~/.fauna-shell
+
+import fs from "fs";
+import os from "os";
+import path from "path";
+const ini = require("ini");
+
+class InvalidConfigError extends Error {}
+
+// Wraps an `ini` file with helpers to get typed values out
+class Config {
+  keyName: string;
+  private config: { [key: string]: unknown };
+
+  static parse(keyName: string, data: string) {
+    return new Config(keyName, ini.parse(data));
+  }
+
+  constructor(keyName: string, config: { [key: string]: unknown }) {
+    this.keyName = keyName;
+    this.config = config;
+  }
+
+  strOpt(key: string): string | undefined {
+    const v = this.config[key];
+    if (v === undefined || typeof v === "string") {
+      return v;
+    } else {
+      throw new InvalidConfigError(
+        `Expected string for ${this.keyName} ${key}, got ${typeof v}`
+      );
+    }
+  }
+
+  numberOpt(key: string): number | undefined {
+    const v = this.config[key];
+    if (v === undefined || typeof v === "number") {
+      return v;
+    } else if (typeof v === "string") {
+      try {
+        return parseInt(v);
+      } catch (_) {
+        throw new InvalidConfigError(
+          `Invalid number for ${this.keyName} ${key}`
+        );
+      }
+    } else {
+      throw new InvalidConfigError(
+        `Expected number for ${this.keyName} ${key}, got ${typeof v}`
+      );
+    }
+  }
+
+  object(key: string): Config {
+    const v = this.config[key];
+    if (v !== null && typeof v === "object") {
+      return new Config(this.keyName, v as any);
```

Comment:
no, because typescript is weird. `as` blindly changes the type of any value, regardless of what typescript thinks. But, the type `object` is not a subtype of the value needed for a config. Here's the error:
```ts
src/lib/config/index.ts:62:39 - error TS2345: Argument of type 'object' is not assignable to parameter of type '{ [key: string]: unknown; }'.
  Index signature for type 'string' is missing in type '{}'.

62       return new Config(this.keyName, v as object);
                                         ~~~~~~~~~~~
```

given that all objects implement the `[]` operator, I don't really understand this error, so for now imma just leave it at `as any`.